tests/components/nest/test_config_flow_sdm.py

```diff
@@ -1,74 +1,225 @@
+
+    entry = get_config_entry(hass)
+    assert entry.title == "Configuration.yaml"
+    assert "token" in entry.data
+    entry.data["token"].pop("expires_at")
+    assert entry.unique_id == DOMAIN
+    assert entry.data["token"] == {
+        "refresh_token": "mock-refresh-token",
+        "access_token": "mock-access-token",
+        "type": "Bearer",
+        "expires_in": 60,
+    }
+
+
+async def test_reauth(hass, oauth):
+    """Test Nest reauthentication."""
+
+    assert await setup.async_setup_component(hass, DOMAIN, CONFIG)
+
+    old_entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={
+            "auth_implementation": DOMAIN,
+            "token": {
+                # Verify this is replaced at end of the test
+                "access_token": "some-revoked-token",
             },
-            "http": {"base_url": "https://example.com"},
+            "sdm": {},
         },
+        unique_id=DOMAIN,
+    )
+    old_entry.add_to_hass(hass)
+
+    entry = get_config_entry(hass)
+    assert entry.data["token"] == {
+        "access_token": "some-revoked-token",
+    }
+
+    await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_REAUTH}, data=old_entry.data
     )
 
+    # Advance through the reauth flow
+    flows = hass.config_entries.flow.async_progress()
+    assert len(flows) == 1
+    assert flows[0]["step_id"] == "reauth_confirm"
+
+    # Run the oauth flow
+    result = await hass.config_entries.flow.async_configure(flows[0]["flow_id"], {})
+    await oauth.async_oauth_flow(result)
+
+    # Verify existing tokens are replaced
+    entry = get_config_entry(hass)
+    entry.data["token"].pop("expires_at")
+    assert entry.unique_id == DOMAIN
+    assert entry.data["token"] == {
+        "refresh_token": "mock-refresh-token",
+        "access_token": "mock-access-token",
+        "type": "Bearer",
+        "expires_in": 60,
+    }
+
+
+async def test_single_config_entry(hass):
+    """Test that only a single config entry is allowed."""
+    old_entry = MockConfigEntry(
+        domain=DOMAIN, data={"auth_implementation": DOMAIN, "sdm": {}}
+    )
+    old_entry.add_to_hass(hass)
+
+    assert await setup.async_setup_component(hass, DOMAIN, CONFIG)
+
     result = await hass.config_entries.flow.async_init(
         DOMAIN, context={"source": config_entries.SOURCE_USER}
     )
-    state = config_entry_oauth2_flow._encode_jwt(
-        hass,
-        {
-            "flow_id": result["flow_id"],
-            "redirect_uri": "https://example.com/auth/external/callback",
-        },
+    assert result["type"] == "abort"
+    assert result["reason"] == "single_instance_allowed"
+
+
+async def test_unexpected_existing_config_entries(hass, oauth):
+    """Test Nest reauthentication with multiple existing config entries."""
+    # Note that this case will not happen in the future since only a single
+    # instance is now allowed, but this may have been allowed in the past.
```

Comment:
What would happen after integration setup in this scenario. Does it run with two config entries loaded?

Reply:
I think when I tested it, it runs the setup twice with all the same configuration.yaml options, then I think it logged a bunch of spam in the error log about having the same devices.

Just to be clear this is not at all expected behavior and not something I expect people to do. 

If this is getting too complicated, happy to just put on it, but then I'd probably still have some gross logic for assuming which entry is correct anyway.

Reply:
Ok, let's keep it like this then.